File: Python/Merge Sorted Array.py
from typing import List


class Solution:
    """
    Time:   O(n+m)
    Memory: O(1)
    """

    def merge(self, first: List[int], n: int, second: List[int], m: int) -> None:
        i, j = n - 1, m - 1

        for curr in reversed(range(n + m)):
            if j < 0:
                first[curr] = first[i]
                i -= 1
            elif i < 0:
                first[curr] = second[j]
                j -= 1
            elif first[i] > second[j]:
                first[curr] = first[i]
                i -= 1
            else:
                first[curr] = second[j]
                j -= 1

File: Java/Merge Sorted Array.java
public class Solution {

    /**
     * Time:   O(n + m)
     * Memory: O(1)
     */
    public void merge(int[] first, int n, int[] second, int m) {
        int i = n - 1, j = m - 1;

        for (int curr = n + m - 1; curr >= 0; --curr)
            if (j < 0)
                first[curr] = first[i--];
            else if (i < 0)
                first[curr] = second[j--];
            else if (first[i] > second[j])
                first[curr] = first[i--];
            else
                first[curr] = second[j--];

    }
}
